### PR TITLE
Add integration test for bluechi-is-online agent with --wait parameter

### LINKED ISSUE #1000: Add integration tests for bluechi-is-online agent with --wait parameter
## Please describe what you would like to see

In https://github.com/eclipse-bluechi/bluechi/issues/962 the new package bluechi-is-online was introduced. However, adding integration tests was postponed. 

A basic integration test for `bluechi-is-online agent` would verify that:
- It exits immediately with code 0 when the `bluechi-agent` is connected
- If bluechi-agent is not connected upon bluechi-is-online execution, bluechi-is-online waits with specified timeout and return 0 if bluechi-agent get connected within this timeout
- If bluechi-agent is not connected upon bluechi-is-online execution, bluechi-is-online waits with specified timeout and return non-zero if bluechi-agent doesn't get connected within this timeout

<br>
<br>

**Tech hints**
- To properly test `--wait` parameter you will need to run `bluechi-is-online agent` in separate thread, for inspiration about using threads in integration tests you can take a look at https://github.com/eclipse-bluechi/bluechi/blob/main/tests/tests/tier0/bluechi-request-cancel/test_bluechi_request_cancel.py

### PR DESCRIPTION
**Pull request by DanaOrr**
Adding integration test for `bluechi-is-online agent --wait` with three tests:
1. Start agent and verify --wait option returns 0 immediately.
2. Stop agent and verify --wait returns 1 after wait time is expired.
3. Stop agent, run 'bluechi-is-online agent --wait' start agent and verify --wait returns 0 before the wait time expires.

Fixes: #1000 
Signed-off-by: Dana Orr [dorr@redhat.com](mailto:dorr@redhat.com)

### CODE REVIEW

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_online,
+            "Checking if agent is online",
+            0,
+            0,
+        ),
+    )
+    check_thread_online.start()
+    check_thread_online.join()
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:67]: Can't we just run `check_agent` in this case, without creating a new thread?

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            result_future_wait,
+            "Checking if agent comes online",
+            20000,
+            0,
+        ),
+    )
+    check_thread_wait.start()
+
+    # Restart the agent after a delay to simulate it coming online during the --wait period
+    time.sleep(5)
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: 5 seconds is quite a lot of time, considering we already approaching the 100 test count, and there will be more of these... It's okay to use sleep when needed, but we should really try to keep our tests as short as possible

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    # wait time expires
+    LOGGER.debug(
+        "Starting test number 3, stopping agent to ensure it is inactive before starting `bluechi-is-online` "
+        "thread."
+    )
+    node_foo.systemctl.stop_unit("bluechi-agent")
+    assert node_foo.wait_for_unit_state_to_be("bluechi-agent", "inactive")
+    LOGGER.debug("Agent confirmed inactive after stopping.")
+
+    # Add a short delay to ensure the "inactive" state is fully processed
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: What do you mean "fully processed"? Isn't `wait_for_unit_state_to_be()` enough?

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    elapsed_time = time.time() - start_time
+    LOGGER.debug(
+        f"Test 3 result: {result_future_wait.result}, Elapsed time: {elapsed_time:.2f} seconds"
+    )
+
+    assert (
+        result_future_wait.result == 0
+    ), "Expected agent to come online before wait expired"
+    assert (
+        elapsed_time < 20
```
Comment (mkemel) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: I would replace it with a timeout that would fail the test after 20 seconds. This way if something goes wrong we would not wait for the external test timeout (which is 45 seconds IIRC).

Also, I would consider a timeout for the other threads. It's not necessary. But maybe you can write some class that runs a thread and also has a timeout and reuse it. Look at `BackgroundRunner` for inspiration.

**Review by mkemel — COMMENTED** [review]
Looks good! 
I have several timing concerns
1. The test has some time.sleeps so that even when everything works fine it runs for quite a long time - 15 seconds. Let's try to make it run faster a bit.
```
            ...
            00:00:03 pass /tests/tier0/bluechi-is-online-agent (on default-0) [31/93]
            00:00:15 pass /tests/tier0/bluechi-is-online-agent-wait (on default-0) [32/93]
            00:00:03 pass /tests/tier0/bluechi-is-online-node (on default-0) [33/93]
            ...
```
2. Please consider having timeout guards for the threads. See comment below

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+
+def check_agent(
+    bluechi_is_online: BluechiIsOnline,
+    future: ResultFuture,
+    description: str,
+    wait_time: int = 0,
+    expected_result: int = 0,
+):
+    result, output = bluechi_is_online.run(
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Could you please adapt existing `agent_is_online` method to accept `wait` parameter when specified and use that method here instead of calling `run` directly? I think `run` method should be internal to `BlueChiIsOnline` class and it shouldn't be invoked from outside ...

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    LOGGER.debug("Starting test number 1 - agent should be online.")
+
+    result_future_online = ResultFuture()
+    check_thread_online = threading.Thread(
+        target=check_agent,
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_online,
+            "Checking if agent is online",
+            0,
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: I'm worried about those hard coded timeouts, we might get into issues when running tests on slower boards.
Could you please expose this value as environment variable (for example `BC_IS_ON_IMMEDIATE_TIMEOUT`) and load it here using fixtures?

You can take a look for details into: https://github.com/eclipse-bluechi/bluechi/pull/970

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            0,
+        ),
+    )
+    check_thread_online.start()
+    check_thread_online.join()
+    assert result_future_online.result == 0, "Expected agent to be online"
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Hmm, shouldn't we also check if the operation ended within expected timeout? Here we expect bluechi-is-online exits immediately, but in following tests we also need to check that it exits within specified timeout.

Hint: please take a look at `Timeout` class and its usage: https://github.com/eclipse-bluechi/bluechi/blob/main/tests/bluechi_test/util.py#L73

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+            node_foo.bluechi_is_online,
+            result_future_offline,
+            "Checking if agent remains offline",
+            5000,
+            1,
+        ),
+    )
+    check_thread_offline.start()
+    check_thread_offline.join()
+    assert result_future_offline.result != 0, "Expected agent to remain offline"
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Here we need to check also that bluechi-is-online really waited specified amount of time before it exited

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+    LOGGER.debug("Starting test number 2 - agent should remain offline.")
+
+    result_future_offline = ResultFuture()
+    check_thread_offline = threading.Thread(
+        target=check_agent,
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_offline,
+            "Checking if agent remains offline",
+            5000,
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Please expose this value as `BC_IS_ON_WAIT_TIMEOUT` using environment variables

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+    LOGGER.debug(
+        "Starting test number 3, stopping agent to ensure it is inactive before starting `bluechi-is-online` "
+        "thread."
+    )
+    node_foo.systemctl.stop_unit("bluechi-agent")
+    assert node_foo.wait_for_unit_state_to_be("bluechi-agent", "inactive")
+    LOGGER.debug("Agent confirmed inactive after stopping.")
+
+    # Add a short delay to ensure the "inactive" state is fully processed
+    time.sleep(1)
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: This could be an issue with slower boards, can't we use `bluechictl` to wait for inactive state?

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    result_future_wait = ResultFuture()
+    start_time = time.time()
+    LOGGER.debug("Starting `bluechi-is-online` thread with wait time of 20 seconds.")
+    check_thread_wait = threading.Thread(
+        target=check_agent,
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_wait,
+            "Checking if agent comes online",
+            20000,
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Please expose this value using env var

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py (hunk 2 above)
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Please expose this value using env var

**Review by mwperina — CHANGES_REQUESTED** [review]

**Comment by mkemel** [comment]
Please rebase instead of merging when updating your repository

File: tests/bluechi_test/bluechi_is_online.py [hunk 12]
```diff
@@ -52,3 +51,12 @@ def system_is_online(self) -> bool:
             0,
         )
         return result == 0
+
+    def agent_is_online_with_wait(self, wait_time: int) -> bool:
```
Comment (mwperina) [inline tests/bluechi_test/bluechi_is_online.py]: Not necessary to have a new method for `wait` parameter, we can manage to have a single method, something like below will allow to have a single method with various input parameters:

```python
 def agent_is_online(self, wait: int = None) -> bool:
        cmd = ["agent"]

        if wait:
            cmd.append("--wait")
            cmd.append(f"{wait}")

        result, output = self.run(
            "Checking if agent is active.",
            " ".join(cmd),
            False,
            0,
        )
        return result == 0
```

File: tests/bluechi_test/fixtures.py [hunk 13]
```diff
@@ -82,6 +82,24 @@ def timeout_test_setup() -> int:
     return _safely_parse_int(_get_env_value("TIMEOUT_TEST_SETUP", ""), 20)
 
 
+@pytest.fixture(scope="session")
+def bc_is_on_wait_timeout() -> int:
+    """Fixture to get the BC_IS_ON_WAIT_TIMEOUT value from environment variable, with a default of 1."""
```
Comment (mwperina) [inline tests/bluechi_test/fixtures.py:87]: Please rename to some meaningful name, for example `BC_IS_ON_IMMEDIATE_TIMEOUT`

File: tests/bluechi_test/fixtures.py [hunk 14]
```diff
@@ -246,10 +264,12 @@ def bluechi_test(
     machines_ssh_user: str,
     machines_ssh_password: str,
     timeout_test_setup: int,
+    bc_is_on_wait_timeout: int,
+    bc_is_on_wait_timeout_5_sec: int,
+    bc_is_on_wait_timeout_20_sec: int,
```
Comment (mwperina) [inline tests/bluechi_test/fixtures.py:269]: I'd not inject those fixtures into a test class, I'd rather inject them directly into `test_bluechi_is_online_agent_wait` method

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    )
+    assert (
+        result_future_wait.result
+    ), "Expected agent to come online before wait expired"
+    assert (
+        2 < elapsed_time < 3
+    ), f"Command took too long to complete: {elapsed_time:.2f} seconds"
+
+
+def test_bluechi_is_online_agent(
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:140]: Please rename the method to match module name: `test_bluechi_is_online_agent_wait`

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py (hunk 1 above)
Comment (mwperina) [reply tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:67]: That makes sense, but we should encapsulate the code with `Timeout` class and check for "immediate" return

File: tests/bluechi_test/bluechi_is_online.py (hunk 12 above)
Comment (DanaOrr) [reply tests/bluechi_test/bluechi_is_online.py]: You are right! Thank you :)

File: tests/bluechi_test/fixtures.py [hunk 16]
```diff
@@ -82,6 +82,24 @@ def timeout_test_setup() -> int:
...
+    """Fixture to get the BC_IS_ON_WAIT_TIMEOUT_5_SEC value from environment variable, with a default of 5."""
+    return _safely_parse_int(_get_env_value("BC_IS_ON_WAIT_TIMEOUT_5_SEC", ""), 5)
+
+
+@pytest.fixture(scope="session")
+def bc_is_on_wait_timeout_20_sec() -> int:
+    """Fixture to get the BC_IS_ON_WAIT_TIMEOUT_20_SEC value from environment variable, with a default of 20."""
+    return _safely_parse_int(_get_env_value("BC_IS_ON_WAIT_TIMEOUT_20_SEC", ""), 20)
+
+
```
Comment (engelmi) [inline tests/bluechi_test/fixtures.py:102]: These fixtures here are just passed to the `bluechi_test` fixture, but never used. In addition, the env-variables are never set as well, so always the default (i.e. here 1, 5, 20) is used. And the defaults used here in the fixtures differ from the defaults in the wait test - e.g. `BC_IS_ON_WAIT_TIMEOUT` is set to 1 in the fixture, but to 1000 in the test. In addition, these timeouts are specific to the currently written test and shouldn't be in here. 
Therefore, please remove the fixtures here (as well as in the `bluechi_test()`) and simply set the timeouts in `test_bluechi_is_online_agent_wait` to the respective values, i.e. `DEFAULT_TIMEOUT=1000`.

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    )
+    future.result = result
+
+
+def exec(ctrl: BluechiControllerMachine, nodes: Dict[str, BluechiAgentMachine]):
+
+    node_foo = nodes[NODE_FOO]
+
+    # Test 1: Start agent and verify --wait option returns 0 immediately
+    node_foo.systemctl.start_unit("bluechi-agent")
```
Comment (engelmi) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:50]: `node-foo` will already be started by the "framework", so you can omit this line - checking for the agent to be active is enough.

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py (hunk 1 above)
Comment (engelmi) [reply tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:67]: Yes, since `bluechi-is-online` should exit immediately and we don't need to do anything in parallel, I'd just wrap this in a timeout - no need for the complexity of threads here. Something like this:
```python
wait_time=0
result = False
start_time = time.time()
with Timeout(wait_time+1, "bluechi-is-online didn't return immediately"):
    result = node_foo.bluechi_is_online(wait_time)
assert result, f"Expected bluechi-is-online with --wait={wait_time} on node {NODE_FOO} to exit true"
assert start_time - time.time() < 1, f"Expected less than 1 second for bluechi-is-online to exit"
```

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+            "Checking if agent remains offline",
+            SHORT_TIMEOUT,
+            1,
+        ),
+    )
+    check_thread_offline.start()
+    check_thread_offline.join()
+    elapsed_time_5_sec = time.time() - start_time_wait_5_sec
+    assert 5 < elapsed_time_5_sec < 6, "--wait returned offline after 5 seconds"
+    assert not result_future_offline.result, "Expected agent to remain offline"
```
Comment (engelmi) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:94]: Actually, this is the same case as @mkemel mentioned above - we don't don anything in parallel and simply wait the `wait_time` interval for `bluechi-is-online` to exit. 
```python
wait_time = 5
result = False
start_time = time.time()
# extend the test timeout here a bit since we are wrapping bluechi-is-online
with Timeout(wait_time + 1, "bluechi-is-online did not exit within the specified time"):
    result = node_foo.bluechi_is_online(wait_time*1000) 
assert not result, f"Expected bluechi-is-online with --wait={wait_time} on node {NODE_FOO} to exit false" 
assert start_time - time.time() <= 5, f"Expected around 5 second for bluechi-is-online to exit"
 
```

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+    )
+    check_thread_wait.start()
+    # Restart the agent after a delay to simulate it coming online during the --wait period
+    time.sleep(SLEEP_DURATION)
+    node_foo.systemctl.start_unit("bluechi-agent")
+    assert node_foo.wait_for_unit_state_to_be("bluechi-agent", "active")
+    LOGGER.debug("Agent confirmed active after starting.")
+
+    # Wait for the thread to complete
+    check_thread_wait.join()
```
Comment (engelmi) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:127]: If the `check_thread_wait` thread hangs - i.e. `bluechi-is-online` keeps running - we'd never stop here. There is already a timeout for the test runs in place, but it is quite high. Since the agent got started the lines before, we can expect the `check_thread_wait.join()` to exit quite soon - so adding a safeguard with ~2s is appropriate:
```python
with Timeout(2, "Thread didn't stop within 2s of starting bluechi-agent"):
    check_thread_wait.join()
```

**Review by engelmi — CHANGES_REQUESTED** [review]

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py (hunk 11 above)
Comment (engelmi) [reply tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: I don't think exposing this value via env var is needed. Its specific to this test, so having it in here or as a constant variable inside this file should be enough. 

Thinking about running the test in multi-host mode: env vars added by a test need to be cleaned up, otherwise it might affect other tests.

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+    assert (
+        2 < elapsed_time < 3
+    ), f"Command took too long to complete: {elapsed_time:.2f} seconds"
+
+
+def test_bluechi_is_online_agent(
+    bluechi_test: BluechiTest,
+    bluechi_node_default_config: BluechiAgentConfig,
+    bluechi_ctrl_default_config: BluechiControllerConfig,
+):
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py:144]: Here I'd inject your fixtures and assign it

```
def test_bluechi_is_online_agent(
    bluechi_test: BluechiTest,
    bluechi_node_default_config: BluechiAgentConfig,
    bluechi_ctrl_default_config: BluechiControllerConfig,
    bc_is_on_wait_timeout: int,
    bc_is_on_wait_timeout_5_sec: int,
    bc_is_on_wait_timeout_20_sec: int,
):

    BC_IS_ON_WAIT_TIMEOUT = bc_is_on_wait_timeout
    BC_IS_ON_WAIT_TIMEOUT_5_sec = bc_is_on_wait_timeout_5_sec
    BC_IS_ON_WAIT_TIMEOUT_20_sec = bc_is_on_wait_timeout_20_sec

```

**Review by mwperina — CHANGES_REQUESTED** [review]

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py (hunk 6 above)
Comment (engelmi) [reply tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: We can't have every timeout specific to tests be configurable by the tmt plan - that would lead to an explosion of timeout configuration. This test alone would add 3+. Adding larger timeouts should be enough. And if the board is really that slow, well...

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py (hunk 11 above)
Comment (mwperina) [reply tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: My idea was to share those timeouts between all bluechi-is-online related tests using `--wait`. And alos the reason why I wanted them to be exposed is that they would need to be manually increased upon test invocation on slower machines

File: tests/bluechi_test/fixtures.py [hunk 21]
```diff
@@ -249,7 +249,6 @@ def bluechi_test(
     timeout_test_run: int,
     timeout_collecting_test_results: int,
 ) -> BluechiTest:
-
     if is_multihost_run:
```
Comment (mwperina) [inline tests/bluechi_test/fixtures.py:273]: Please undo this change

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+    future.result = result
+
+
+def exec(ctrl: BluechiControllerMachine, nodes: Dict[str, BluechiAgentMachine]):
+
+    node_foo = nodes[NODE_FOO]
+
+    # Test 1: Start agent and verify --wait option returns 0 immediately
+    try:
+        with Timeout(DEFAULT_WAIT_TIMEOUT + 1, "Timeout during Test 1"):
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Why do you need to increment `DEAFULT_WAIT_TIMEOUT`?

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+from bluechi_test.bluechi_is_online import BluechiIsOnline
+from bluechi_test.config import BluechiAgentConfig, BluechiControllerConfig
+from bluechi_test.machine import BluechiAgentMachine, BluechiControllerMachine
+from bluechi_test.test import BluechiTest
+from bluechi_test.util import Timeout
+
+LOGGER = logging.getLogger(__name__)
+
+NODE_FOO = "node-foo"
+DEFAULT_WAIT_TIMEOUT = int(os.getenv("BC_IS_ON_WAIT_TIMEOUT", 1000))
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Hmm, shouldn't this be 1 second and not 1000 seconds?
Also please rename to some meaningful name, for example ` IMMEDIATE_RETURN_TIMEOUT`

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+SHORT_WAIT_TIMEOUT = int(os.getenv("BC_IS_ON_WAIT_TIMEOUT_SHORT", 5000))
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: How about `WAIT_PARAM_VALUE`?

File: tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py [hunk 25]
```diff
@@ -0,0 +1,163 @@
...
+                wait_time=DEFAULT_WAIT_TIMEOUT,
+                expected_result=0,
+            )
+            elapsed_time = time.time() - start_time_immediately
+            assert (
+                elapsed_time < 1 and result_future_online.result
+            ), "Expected less than 1 second and the agent to be online"
+    except TimeoutError as ex:
+        LOGGER.error(f"Test 1 failed: {ex}")
+        return
```
Comment (mwperina) [inline tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py]: Can't you simplify the whole block to just below?

```python
with Timeout(IMMEDIATE_RETURN_TIMEOUT, "bluechi-is-online didn't return immediately"):
    assert node_foo.bluechi_is_online.agent_is_online(wait=WAIT_PARAM_VALUE) == 0
)
```

File: tests/bluechi_test/fixtures.py
Comment (mwperina) [inline tests/bluechi_test/fixtures.py:1]: Please undo this change
